libraries/Microsoft.Bot.Connector/Authentication/AppCredentials.cs

```diff
@@ -0,0 +1,232 @@
+            }
+
+            set
+            {
+                // Advanced user only, see https://aka.ms/bots/tenant-restriction
+                var endpointUrl = string.Format(AuthenticationConstants.ToChannelFromBotLoginUrlTemplate, value);
+
+                if (Uri.TryCreate(endpointUrl, UriKind.Absolute, out var result))
+                {
+                    AuthTenant = value;
+                }
+                else
+                {
+                    throw new Exception($"Invalid channel auth tenant: {value}");
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets the OAuth endpoint to use.
+        /// </summary>
+        /// <value>
+        /// The OAuth endpoint to use.
+        /// </value>
+        public virtual string OAuthEndpoint => string.Format(AuthenticationConstants.ToChannelFromBotLoginUrlTemplate, ChannelAuthTenant);
+
+        /// <summary>
+        /// Gets the OAuth scope to use.
+        /// </summary>
+        /// <value>
+        /// The OAuth scope to use.
+        /// </value>
+        public virtual string OAuthScope => AuthenticationConstants.ToChannelFromBotOAuthScope;
+
+        /// <summary>
+        /// Gets or sets the channel auth token tenant for this credential.
+        /// </summary>
+        /// <value>
+        /// The channel auth token tenant for this credential.
+        /// </value>
+        protected string AuthTenant { get; set; }
+
+        /// <summary>
+        /// Gets or sets the channel auth token tenant for this credential.
+        /// </summary>
+        /// <value>
+        /// The channel auth token tenant for this credential.
+        /// </value>
+        protected HttpClient CustomHttpClient { get; set; }
+
+        /// <summary>
+        /// Gets or sets the channel auth token tenant for this credential.
+        /// </summary>
+        /// <value>
+        /// The channel auth token tenant for this credential.
+        /// </value>
+        protected ILogger Logger { get; set; }
+
+        /// <summary>
+        /// Adds the host of service url to <see cref="MicrosoftAppCredentials"/> trusted hosts.
+        /// </summary>
+        /// <param name="serviceUrl">The service URL.</param>
+        /// <remarks>If expiration time is not provided, the expiration time will DateTime.UtcNow.AddDays(1).</remarks>
+        public static void TrustServiceUrl(string serviceUrl)
+        {
+            TrustServiceUrl(serviceUrl, DateTime.UtcNow.Add(TimeSpan.FromDays(1)));
+        }
+
+        /// <summary>
+        /// Adds the host of service url to <see cref="MicrosoftAppCredentials"/> trusted hosts.
+        /// </summary>
+        /// <param name="serviceUrl">The service URL.</param>
+        /// <param name="expirationTime">The expiration time after which this service url is not trusted anymore.</param>
+        public static void TrustServiceUrl(string serviceUrl, DateTime expirationTime)
+        {
+            lock (TrustedHostNames)
```

Comment:
if the dictionary TrustedHostNames were a Concurrent Dictionary, could we eliminate these locks?